internal/cmd/update.go

```diff
@@ -0,0 +1,466 @@
+			ctx := cmd.Context()
+			httpClient := &http.Client{
+				Timeout: 10 * time.Second,
+			}
+
+			currExe, err := os.Executable()
+			if err != nil {
+				return clog.Fatal("init: get current executable", clog.Causef(err.Error()))
+			}
+
+			updater := &updater{
+				confirmF:       defaultConfirm,
+				execF:          defaultExec,
+				executablePath: currExe,
+				httpClient:     httpClient,
+				fs:             afero.NewOsFs(),
+				osF:            func() string { return runtime.GOOS },
+				versionF:       func() string { return version.Version },
+			}
+			return updater.Run(ctx, force, coderURL, versionArg)
+		},
+	}
+
+	cmd.Flags().BoolVar(&force, "force", false, "do not prompt for confirmation")
+	cmd.Flags().StringVar(&coderURL, "coder", "", "query this coder instance for the matching version")
+	cmd.Flags().StringVar(&versionArg, "version", "", "explicitly specify which version to fetch and install")
+
+	return cmd
+}
+
+type getter interface {
+	Get(url string) (*http.Response, error)
+}
+
+func (u *updater) Run(ctx context.Context, force bool, coderURLArg string, versionArg string) error {
+	// Check under following directories and warn if coder binary is under them:
+	//   * C:\Windows\
+	//   * homebrew prefix
+	//   * coder assets root (/var/tmp/coder)
+	var pathBlockList = []string{
+		`C:\Windows\`,
+		`/var/tmp/coder`,
+	}
+	brewPrefixCmd, err := u.execF(ctx, "brew", "--prefix")
+	if err == nil { // ignore errors if homebrew not installed
+		pathBlockList = append(pathBlockList, strings.TrimSpace(string(brewPrefixCmd)))
+	}
+
+	for _, prefix := range pathBlockList {
+		if HasFilePathPrefix(u.executablePath, prefix) {
+			return clog.Fatal(
+				"cowardly refusing to update coder binary",
+				clog.BlankLine,
+				clog.Causef("executable path %q is under blocklisted prefix %q", u.executablePath, prefix))
+		}
+	}
+
+	currentBinaryStat, err := u.fs.Stat(u.executablePath)
+	if err != nil {
+		return clog.Fatal("preflight: cannot stat current binary", clog.Causef(err.Error()))
+	}
+
+	if currentBinaryStat.Mode().Perm()&0222 == 0 {
+		return clog.Fatal("preflight: missing write permission on current binary")
+	}
+
+	clog.LogInfo(fmt.Sprintf("Current version of coder-cli is %s", version.Version))
+
+	desiredVersion, err := getDesiredVersion(u.httpClient, coderURLArg, versionArg)
+	if err != nil {
+		return clog.Fatal("failed to determine desired version of coder", clog.Causef(err.Error()))
+	}
+
+	currentVersion, err := semver.StrictNewVersion(u.versionF())
+	if err != nil {
+		clog.LogWarn("failed to determine current version of coder-cli", clog.Causef(err.Error()))
+	} else if currentVersion.Compare(desiredVersion) == 0 {
+		clog.LogInfo("Up to date!")
+		return nil
+	}
+
+	if !force {
+		label := fmt.Sprintf("Do you want to download version %s instead", desiredVersion)
+		if _, err := u.confirmF(label); err != nil {
+			return clog.Fatal("failed to confirm update", clog.Tipf(`use "--force" to update without confirmation`))
```

Comment:
instead of "failed to confirm update", I think "user cancelled operation" or similar would be clearer?